compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2349,6 +2349,14 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     None
                 }
             }
+            GenericKind::Opaque(def_id, _substs) => {
+                // Avoid emitting a `... so that the type` message at the error site.
+                // It would be out of order for return position impl trait
+                origin = None;
+                // Make sure the lifetime suggestion is on the RPIT instead of proposing
+                // to add a bound for opaque types (which isn't possible)
+                Some((self.tcx.def_span(def_id).shrink_to_hi(), true))
```

Comment:
This would suggest making changes at the opaque type's definition site and the suggested lifetime probably cannot be named in that scope.

I still think removing the `need_infer` hack for opaque types in https://github.com/rust-lang/rust/pull/95474/files#r880479741 is the way to go. After all it's just a diagnostics hack for projection types only.

for example:
```rust
// Now we suggest adding `+ 'static` to the RPIT here.
fn opaque<T>(_: T) -> impl Sized {}

// ... but I think we should suggest `T: 'static` here similar to the current stable.
fn test<T>(t: T) -> impl Sized + 'static {
    opaque(t)
}
```